File: LeetCode/15.三数之和.go
import "sort"

/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	result := [][]int{}
	if len(nums) < 1 {
		return result
	}
	sort.Ints(nums)

	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		p1 := i + 1
		p2 := len(nums) - 1
		for p1 < p2 {
			if nums[i]+nums[p1]+nums[p2] == 0 {
				result = append(result, []int{nums[i], nums[p1], nums[p2]})
				// 注意在循环中对循环判断变量做修改，要首先满足循环条件
				for p1 < p2 && nums[p1+1] == nums[p1] {
					p1++
				}
				for p1 < p2 && nums[p2-1] == nums[p2] {
					p2--
				}
				p1++
				p2--
			} else if nums[p1]+nums[p2] < -nums[i] {
				p1++
			} else {
				p2--
			}

		}
	}
	return result
}

// @lc code=end

File: LeetCode/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
#include<algorithm>
#include<vector>
#include<set>
#include<iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        return sortFind(nums);
    }


    // O(n^2)
    vector<vector<int>> sortFind(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        set<int> sumRecord;
        for(int i = 0; i < nums.size()-2; i++) {
            // 跳过连续的相同数字(刚处理完结果一样，一方面节省失败验证的时间，一方面避免重复结果)
            // 依赖前一步排序
            if(i >= 1 && nums[i] == nums[i-1]) {
                continue;
            }

            // 内层循环 寻找第二个数并处记录两数和
            for(int j = i+1; j < nums.size(); j++) {
                if(sumRecord.find(-nums[j]-nums[i])==sumRecord.end()) {
                    sumRecord.insert(-nums[i]-nums[j]);
                } else {
                    res.push_back(vector<int> {-nums[i] - nums[j], nums[i], nums[j]});
                }
            }
        }
        return res;
    }
};

int main(void) {
    vector<int> nums = {-1,0,1,2,-1,-4};
    Solution s;
    vector<vector<int>> res = s.threeSum(nums);
    return 0;
}
// @lc code=end

